core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java

```diff
@@ -2354,6 +2377,33 @@ boolean hasAliasUsedInGroupByWhichIsNotPresentInFinalProjection(Project rel) {
       return false;
     }
 
+    boolean hasFieldsUsedInFilterWhichIsNotUsedInFinalProjection(Project project) {
+      final SqlNodeList selectList = ((SqlSelect) node).getSelectList();
```

Comment:
can we simplify this  and take it inside variable.

SqlSelect sqlSelect = (SqlSelect) node;
    SqlNodeList selectList = **sqlSelect**.getSelectList();
    List<SqlNode> qualifyColumnList = new ArrayList<>();

    **sqlSelect**.getQualify()